File: SiderShopping/app/src/main/java/com/mmart/sidershopping/model/api/RetrofitClient.java
package com.mmart.sidershopping.model.api;

import com.mmart.sidershopping.model.ALodjinhaAPI;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitClient {

    private static final String BASE_URL = "https://alodjinha.herokuapp.com";
    private static RetrofitClient mInstace;
    private Retrofit retrofit;


    private RetrofitClient() {

        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient client = new OkHttpClient.Builder().addInterceptor(interceptor).build();

        retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build();


    }


    public static synchronized RetrofitClient getInstance() {
        if (mInstace == null) {
            mInstace = new RetrofitClient();
        }

        return mInstace;
    }

    public ALodjinhaAPI api() {

        return retrofit.create(ALodjinhaAPI.class);
    }
}

File: aLodjinha/app/src/main/java/br/com/andreguedes/alodjinha/data/source/RetrofitClient.kt
package br.com.andreguedes.alodjinha.data.source

import br.com.andreguedes.alodjinha.BuildConfig
import br.com.andreguedes.alodjinha.ALodjinhaApplication
import br.com.andreguedes.alodjinha.data.source.remote.ALodjinhaAPI
import okhttp3.Cache
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.Retrofit
import okhttp3.OkHttpClient
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import java.util.concurrent.TimeUnit

object RetrofitClient {

    private var retrofit: Retrofit? = null

    fun getClient(): ALodjinhaAPI {
        val cache = Cache(ALodjinhaApplication.getFileCacheDir(), 10 * 1024 * 1024)

        val client = OkHttpClient.Builder()
                .readTimeout(20, TimeUnit.SECONDS)
                .connectTimeout(20, TimeUnit.SECONDS)
                .cache(cache)
                .addInterceptor { chain ->
                    var request = chain.request()
                    request = if (ALodjinhaApplication.isNetworkAvailable()) {
                        request.newBuilder().header("Cache-Control", "public, max-age=" + 60).build()
                    } else {
                        request.newBuilder().header("Cache-Control", "public, only-if-cached, max-stale=" + 60 * 60 * 24 * 7).build()
                    }
                    chain.proceed(request)
                }
                .build()

        if (retrofit == null) {
            retrofit = Retrofit.Builder()
                    .baseUrl(BuildConfig.END_POINT)
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                    .client(client)
                    .build()
        }
        return retrofit!!.create(ALodjinhaAPI::class.java)
    }

}